public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx

```diff
@@ -0,0 +1,84 @@
+import { SceneObjectBase, SceneObjectState, VizPanel, VizPanelState } from '@grafana/scenes';
+import { LibraryPanel } from '@grafana/schema';
+import { PanelModel } from 'app/features/dashboard/state';
+import { getLibraryPanel } from 'app/features/library-panels/state/api';
+
+import { createPanelDataProvider } from '../utils/createPanelDataProvider';
+
+import { DashboardGridItem } from './DashboardGridItem';
+
+interface LibraryPanelBehaviorState extends SceneObjectState {
+  // Library panels use title from dashboard JSON's panel model, not from library panel definition, hence we pass it.
+  title?: string;
+  uid: string;
+  name: string;
+  isLoaded?: boolean;
+  _loadedPanel?: LibraryPanel;
+}
+
+export class LibraryPanelBehavior extends SceneObjectBase<LibraryPanelBehaviorState> {
+  public constructor(state: LibraryPanelBehaviorState) {
+    super(state);
+
+    this.addActivationHandler(() => this._activationHandler());
+  }
+
+  private _activationHandler() {
+    if (!this.state.isLoaded) {
+      this.loadLibraryPanelFromPanelModel();
+    }
+  }
+
+  public setPanelFromLibPanel(libPanel: LibraryPanel) {
+    if (this.state._loadedPanel?.version === libPanel.version) {
+      return;
+    }
+
+    const vizPanel = this.parent;
+
+    if (!(vizPanel instanceof VizPanel)) {
+      return;
+    }
+
+    const gridItem = vizPanel.parent;
+
+    if (!(gridItem instanceof DashboardGridItem)) {
+      return;
+    }
+
+    const libPanelModel = new PanelModel(libPanel.model);
+
+    const vizPanelState: VizPanelState = {
+      title: libPanelModel.title,
+      options: libPanelModel.options ?? {},
+      fieldConfig: libPanelModel.fieldConfig,
+      pluginId: libPanelModel.type,
+      pluginVersion: libPanelModel.pluginVersion,
+      displayMode: libPanelModel.transparent ? 'transparent' : undefined,
+      description: libPanelModel.description,
+      $data: createPanelDataProvider(libPanelModel),
+    };
+
+    this.setState({ _loadedPanel: libPanel, isLoaded: true, name: libPanel.name, title: libPanelModel.title });
+
+    const clone = vizPanel.clone(vizPanelState);
+    gridItem.setState({ body: clone });
```

Comment:
My biggest problem with this behavior approach which I discovered fixing issues in https://github.com/grafana/grafana/pull/92846/ 

is this setState. Feels a bit broken that the behavior replaces it's parent instance (and clones itself), by going to it's parent's parent (the DashboardGridItem) and updates it's state. 

I think we could get it work without updating the grid item state and creating a VizPanel instance

if we did: 
```ts
    vizPanel.setState(vizPanelState);
    vizPanel.changePluginType(libPanelModel.type, vizPanelState.options, vizPanelState.fieldConfig);
```

It seems to work. For panel edit we would still need to wait (as doing this after we open panel edit could have issues, but might work)

Reply:
I explored doing that and could then get much simpler fix for the view panel / edit panel (from full page reload)

https://github.com/grafana/grafana/pull/92850

Reply:
@torkelo I just went through this again, just to make sure I understand: What isn't working properly in this PR on page reload in edit/view panel? I just retested this and it loads in both view/edit on page reload. What am I missing?

Reply:
@mdvictor eh, that is super strange, a simple lib panel added to a dashboard, then save. go into view panel, do a full page reload, see not correct queries not viz. 

Same issue in panel edit, and the explanation is super easy to understand in view panel and panel edit the VizPanel parent is not g grid item so no replacement of the VizPanel with the loaded lib panel happens.

[testing---Dashboards---Grafana.webm](https://github.com/user-attachments/assets/95936e36-bf8a-40f3-9739-db0733e0dee5)

Reply:
@mdvictor I think you must have tested with a repeated lib panel, and the reason it works there is that in that case we wait for the repeat process to complete before going into view / edit, and that wait gives the behavior time to load and replace the viz panel (but depends on timing of the load lib panel API call I guess, if it was slow we would still see bug there)

but just test with a lib panel that is not repeated

Reply:
Ahh yes, I see it now, thank you!